sklearn/cluster/affinity_propagation_.py

```diff
@@ -327,4 +361,10 @@ def predict(self, X):
             raise ValueError("Predict method is not supported when "
                              "affinity='precomputed'.")
 
-        return pairwise_distances_argmin(X, self.cluster_centers_)
+        if self.cluster_centers_.size > 0:
+            return pairwise_distances_argmin(X, self.cluster_centers_)
+        else:
+            warnings.warn("This model does not have any cluster centers "
+                          "because affinity propagation did not converge. "
+                          "Returning unique labels for the provided samples.")
```

Comment:
No longer the case.

Are you sure we don't just want to raise an error when a user tries to predict with such a model? Perhaps an array of -1 makes sense in fit_predict if we're going to do it here..?

Reply:
`predict()` unexpectedly raising an error was the initial reason this PR got started. An error during `predict()` would be appropriate if the caller would provide incorrect data IMHO. Since these models can potentially live for a long time (in my use case, they're trained infrequently and are deserialized later with `joblib.load()`), I wouldn't want to make the caller of `predict()` responsible for dealing with potential crashes because of issues during training of the models. Returning -1 as labels (and logging the warning) seems to be a bit more friendly to the caller.

I'll address the incorrect warning message.